### PR TITLE
add VK_KHR_maintenance5 support

### LINKED ISSUE #3244: support for VK_KHR_maintenance5
## Description

Renderdoc currently suppresses the `VK_KHR_maintenance5` Vulkan extension. I cannot see why such limitation is here (probably captured extensions were not recently updated) and would expect its support.

## Environment

* RenderDoc version: v1.31
* Operating System: Ubuntu 23.10
* Graphics API: Vulkan

### PR DESCRIPTION
**Pull request by qbojj**
## Description

This PR adds `VK_KHR_maintenance5` support. (resolve #3244)

Spec: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_maintenance5.html
Proposal: https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/VK_KHR_maintenance5.adoc

Here are features that are added in `VK_KHR_maintenance5` and this PR should support (features that I am unsure of are not set):

- [x] A new VK_FORMAT_A1B5G5R5_UNORM_PACK16_KHR format
- [x] A new VK_FORMAT_A8_UNORM_KHR format
- [x] A property to indicate that multisample coverage operations are performed after sample counting in EarlyFragmentTests mode
- [x] Relax VkBufferView creation requirements by allowing subsets of the associated VkBuffer usage using VkBufferUsageFlags2CreateInfoKHR
- [x] A new entry point [vkCmdBindIndexBuffer2KHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdBindIndexBuffer2KHR.html), allowing a range of memory to be bound as an index buffer
- [x] [vkGetDeviceProcAddr](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceProcAddr.html) must return NULL for supported core functions beyond the version requested by the application.
- [x] A property to indicate that the sample mask test is performed after sample counting in EarlyFragmentTests mode
- [x] vkCmdBindVertexBuffers2 now supports using VK_WHOLE_SIZE in the pSizes parameter.
- [x] A default size of 1.0 is used if PointSize is not written
- [x] Shader modules are deprecated - applications can now pass [VkShaderModuleCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkShaderModuleCreateInfo.html) as a chained struct to pipeline creation via [VkPipelineShaderStageCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html)
- [x] A function [vkGetRenderingAreaGranularityKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetRenderingAreaGranularityKHR.html) to query the optimal render area for a dynamic rendering instance.
- [x] A property to indicate that depth/stencil texturing operations with VK_COMPONENT_SWIZZLE_ONE have defined behavior
- [x] Add [vkGetImageSubresourceLayout2KHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetImageSubresourceLayout2KHR.html) and a new function [vkGetDeviceImageSubresourceLayoutKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceImageSubresourceLayoutKHR.html) to allow the application to query the image memory layout without having to create an image object and query it.
- [x] Allow VK_REMAINING_ARRAY_LAYERS as the layerCount member of [VkImageSubresourceLayers](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkImageSubresourceLayers.html)
- [x] Adds stronger guarantees for propagation of VK_ERROR_DEVICE_LOST return values
- [x] A property to indicate whether PointSize controls the final rasterization of polygons if [polygon mode](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#primsrast-polygonmode) is VK_POLYGON_MODE_POINT
- [x] Two properties to indicate the non-strict line rasterization algorithm used
- [x] Two new flags words [VkPipelineCreateFlagBits2KHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineCreateFlagBits2KHR.html) and [VkBufferUsageFlagBits2KHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkBufferUsageFlagBits2KHR.html)
- [x] Physical-device-level functions can now be called with any value in the valid range for a type beyond the defined enumerants, such that applications can avoid checking individual features, extensions, or versions before querying supported properties of a particular enumerant.
- [x] Clarification that copies between images of any type are allowed, treating 1D images as 2D images with a height of 1.

### CODE REVIEW

**Comment by qbojj** [comment]
There is some things I do not know where to implement:
- interaction of robustness and index buffer (size is not VK_WHOLE_SIZE):
  - mesh viewer should show index=0 if robustness is enabled and is dereferencing beyond bound buffer
- default size of PointSize (didn't look at shader debugging yet)

Also the point
> [vkGetDeviceProcAddr](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceProcAddr.html) must return NULL for supported core functions beyond the version requested by the application.

is implemented as somewhat of a hack (though it makes this feature zero-maintainable).

I could change it to make version annotations in the `HookInitVulkanDevice` macro.

**Comment by baldurk** [comment]
Can you clarify how you want to handle this PR?

Do you want this partial support as-is reviewed and merged while you investigate the remaining issues, and then you'll PR those and enable the extension? Or are you wanting to continue working on this and complete it before you get a review?

**Comment by qbojj** [comment]
The only feature left to implement is non-whole `VkBuffer` index buffers, which only surfaces when robustness is enabled. And personally I do not use robustness. I could add the feature myself but I have no idea where is the mesh view implemented.

So basically I am happy with where this PR is currently (you can merge it now), but if You could point me into correct direction I could add the robustness on index buffers myself.

**Comment by qbojj** [comment]
Do not merge yet. I am currently testing.

**Comment by qbojj** [comment]
I have finished all parts of this extension, so this PR is ready for review.

**Comment by qbojj** [comment]
It looks like setting PointSize to 1.0f where I tried doesn't work... IDK how to do it

**Comment by qbojj** [comment]
I would like it merged even if partially.

**Comment by qbojj** [comment]
Default 1.0f for PointSize now works (everything is implemented).

**Comment by baldurk** [comment]
Just checking - is this PR ready for review? I've noticed you pushed and commented several times since you said it was complete. If you'd rather wait if you're still working on it, please mark it as draft or close and re-open it.

**Comment by qbojj** [comment]
Yes, those were some code style changes. Sorry for the confusion.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp [hunk 1]
```diff
@@ -41,6 +41,13 @@ Q_DECLARE_METATYPE(CombinedSamplerData);
 
 namespace
 {
+
+template <typename T>
+T RDCMIN(const T &a, const T &b)
+{
+  return a < b ? a : b;
+}
+
```
Comment (baldurk) [inline qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp]: Please don't define new macros in qrenderdoc like this. The Qt code uses `qMin` and `qMax` as defined by Qt to fit that project.

File: renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp [hunk 2]
```diff
@@ -966,6 +966,15 @@ ShaderDebugTrace *Debugger::BeginDebug(DebugAPIWrapper *api, const ShaderStage s
...
         }
 
+        // VK_KHR_maintenance5 requires that if PointSize is not written during shader execution, it should be 1.0f
+        //   so we will initialize it to 1.0f here (if it is not enabled then leaving it as 1.0f is harmless)
+        if(!isInput && curDecorations.flags & Decorations::HasBuiltIn &&
+           curDecorations.builtIn == BuiltIn::PointSize)
+        {
+          RDCASSERTEQUAL(var.type, VarType::Float);
+          var.value.f32v[0] = 1.0f;
+        }
```
Comment (baldurk) [inline renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp]: What's the motivation for this change? The behaviour difference in `VK_KHR_maintenance5` is effectively considered _after_ the shader has executed, if `PointSize` is not written to. I don't think that means we should pretend that it's initialised to that value as that feels misleading.

File: renderdoc/driver/vulkan/vk_core.h [hunk 3]
```diff
@@ -2812,4 +2812,18 @@ class WrappedVulkan : public IFrameCapturer
                                 VkCommandBuffer commandBuffer, VkBuffer buffer, VkDeviceSize offset,
                                 VkBuffer countBuffer, VkDeviceSize countBufferOffset,
                                 uint32_t maxDrawCount, uint32_t stride);
+
+  // VK_KHR_mainenance5
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_core.h]: nit: typo
```suggestion
  // VK_KHR_maintenance5
```

File: renderdoc/driver/vulkan/vk_layer.cpp [hunk 4]
```diff
@@ -413,6 +413,16 @@ VK_LAYER_RENDERDOC_CaptureEnumerateInstanceExtensionProperties(
...
 {
+  if(device == VK_NULL_HANDLE || pName == NULL)
+    return NULL;
+
+  PFN_vkGetDeviceProcAddr gpa = GetDeviceDispatchTable(Unwrap(device))->GetDeviceProcAddr;
+
+  // if GPA returns NULL, that means that the standard prohibits non-NULL values for this pName
+  //   or that the return value is undefined
+  if(!gpa || !gpa(Unwrap(device), pName))
+    return NULL;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_layer.cpp]: This change seems wrong to me as we should not rely on upstream layers/loader to define our correctness. Aside from that the loader interface document says that layers *must not* call down the chain inside `vkGetDeviceProcAddr`.

What is the problem you're trying to solve here? The behaviour difference defined in `KHR_maintenance5` should already be correctly handled by RenderDoc.

File: renderdoc/driver/vulkan/vk_overlay.cpp [hunk 5]
```diff
@@ -364,9 +364,19 @@ void VulkanDebugManager::PatchLineStripIndexBuffer(const ActionDescription *acti
...
+
+    uint64_t len = RDCMIN(data_len, rs.ibuffer.size - offset);
+
+    GetBufferData(rs.ibuffer.buf, rs.ibuffer.offs + offset, len, indices);
+
+    if(len < data_len)
+    {
+      // fill the rest with 0s
+      indices.resize(static_cast<size_t>(data_len));
+      memset(indices.data() + len, 0, static_cast<size_t>(data_len - len));
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_overlay.cpp]: Please prefer using normal casts, not `static_cast`.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 6]
```diff
@@ -3893,6 +3893,94 @@ void WrappedVulkan::vkCmdBindIndexBuffer(VkCommandBuffer commandBuffer, VkBuffer
...
+      {
+        commandBuffer = RerecordCmdBuf(m_LastCmdBufferID);
+        ObjDisp(commandBuffer)
+            ->CmdBindIndexBuffer2KHR(Unwrap(commandBuffer), Unwrap(buffer), offset, size, indexType);
+
+        {
+          VulkanRenderState &renderstate = GetCmdRenderState();
+          renderstate.ibuffer.buf = GetResID(buffer);
+          renderstate.ibuffer.offs = offset;
+          renderstate.ibuffer.size = size;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:3943]: The normal call to `vkCmdBindIndexBuffer` needs to set the new size value to `VK_WHOLE_SIZE`, otherwise any size bound here will be leftover and incorrectly apply to subsequent non-sized index buffer binds.

**Review by baldurk — CHANGES_REQUESTED** [review]
I've left some notes on some things, but this PR also seems to be missing some parts - at least one is crucial.

There's no handling of the new structures defined in the extension, the standard features struct but also the new extended usage and pipeline create flags structs which must be used in preference to the normal enums if present. These structs aren't serialised or handled in next chains. I would also expect to see new handling for the new format enums.

Most importantly though the extension isn't whitelisted, so no application will be able to use it. Maybe that's deliberate if this is intended to be unfinished/unused, but you said the PR and extension support was complete so I'm assuming it's an omission.

On the last point, how did you test this code and the extension? Did you allow use of it locally and then remove that for this PR? I'm a little unclear on what the status of this is still.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp (hunk 6 above)
Comment (qbojj) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:3943]: I have set this member to be defaulted to `VK_WHOLE_SIZE`

File: renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp (hunk 2 above)
Comment (qbojj) [reply renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp]: OK, will revert

**Comment by qbojj** [comment]
> There's no handling of the new structures defined in the extension, the standard features struct but also the new extended usage and pipeline create flags structs which must be used in preference to the normal enums if present. These structs aren't serialised or handled in next chains. I would also expect to see new handling for the new format enums.

Oh... as I am not using this part of the extension I hadn't noticed the lack of support. I will look into it.

> Most importantly though the extension isn't whitelisted, so no application will be able to use it. Maybe that's deliberate if this is intended to be unfinished/unused, but you said the PR and extension support was complete so I'm assuming it's an omission.
>
> On the last point, how did you test this code and the extension? Did you allow use of it locally and then remove that for this PR? I'm a little unclear on what the status of this is still.

Strange. When I build it locally and run my program that enables this extension it works just fine (it enumerates the extension and enabling it works). I do test it with my program, but It doesn't use all of the features. I should probably make a test program that uses everything from this extension.

Anyways I have changed the parts You have commented upon but am yet to add support for new usage/flags, so when that will be done I will message You.

File: renderdoc/driver/vulkan/vk_layer.cpp (hunk 4 above)
Comment (qbojj) [reply renderdoc/driver/vulkan/vk_layer.cpp]: changed it so functions explicitly say in which version of vulkan they were integrated into core

File: renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp (hunk 2 above)
Comment (baldurk) [reply renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp]: I do think it's not necessary but I would like to know the reason you had for adding the code originally. Maybe it's something I hadn't considered or maybe you're trying to solve a problem which needs to be solved a different way.

File: renderdoc/driver/vulkan/vk_layer.cpp (hunk 4 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_layer.cpp]: Version gating is already handled per extension by the condition, so I still don't see the purpose of this code change.

I asked above what the problem is that you're trying to solve, and you've not answered that question. Changing the code without answering the question or explaining yourself doesn't address this feedback.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp (hunk 6 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:3943]: I saw that, but that does not solve the problem. If an application calls `vkCmdBindIndexBuffer2KHR` with a size specified of 128 bytes, and then later in the same command buffer calls `vkCmdBindIndexBuffer` which does not provide a size in the function call the renderstate size will not be updated so the new binding will still contain a size of 128 bytes, even if the buffer is larger and it should be accessed with an unclamped size.

**Comment by baldurk** [comment]
> Strange. When I build it locally and run my program that enables this extension it works just fine (it enumerates the extension and enabling it works). I do test it with my program, but It doesn't use all of the features. I should probably make a test program that uses everything from this extension.

I don't see any way that that is possible, unless you have local changes to RenderDoc that are not present in the PR here. Not only will RenderDoc only enumerate supported device extensions, but at `vkCreateDevice` time RenderDoc will check to see that all enabled extensions are supported and it will fail device creation if an unsupported device extension is passed.

The only possibilities I can see is that your application does not hard-require KHR_maintenance5 and when run under RenderDoc it is running the same as it would on any driver that didn't support the extension, or else it is not written legally and is relying on/using functionality from the extension without either enabling it or passing the feature struct.

**Comment by qbojj** [comment]
my bad. I forgot to uncomment the extension string, but it was still passing the features struct to create device. Now I have added it to whitelisted extensions.

File: renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp (hunk 2 above)
Comment (qbojj) [reply renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp]: My train of thought was that I should mark that the pointSize has such behaviour. But now that I am thinking about it if someone is seeing that PointSize turns out to be 1. (as side effects) even if he hadn't written to it is better than having to debug if this ext is not enabled why PointSize is shown in debugger to be 1. but they are rendered with garbage PointSize.

File: renderdoc/driver/vulkan/vk_layer.cpp (hunk 4 above)
Comment (qbojj) [reply renderdoc/driver/vulkan/vk_layer.cpp]: I'm sorry. It is just if for example `VK_KHR_dynamic_rendering` is enabled with requested instance version of 1.2. RenderDoc will resolve `vkCmdBeginRendering` but it was not core in version 1.3 so it should be NULL. (https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceProcAddr.html 2nd annotation of Table 1.)

**Comment by qbojj** [comment]
I have since tested this PR and found one problem with it (I am not sure how should I fix those):

https://github.com/qbojj/vk_khr_maintenance5_test/tree/master
When `robustBufferAccess2` is enabled and index is used outside of bound region the mesh viewer shows a `---` at the IDX column (this is expected), but the shader input columns are also written as `---` (those should be pulled from `index=0`). Shader debugger pulls data as expected.

Also https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceProcAddr.html states that it should resolve (for a valid device handle) only for: "requested core version device-level dispatchable commands" and
"enabled extension device-level dispatchable commands". I have handled the first part, but from my understanding the second part doesn't include promoted extensions from core (only explicitly enabled extensions should be reported) (that it what mesa/radv follows).

On the other hand it looks like vertex buffer robustness is not working (even when not using it on the index buffer so unrelated to this PR):
  https://github.com/qbojj/vk_khr_maintenance5_test/tree/bad-robustness
The state of the command `vkBindVertexBuffers2(EXT)` is not properly restored during the replay (see the call to vkDrawIndexed). Also when debugging the vertex shader instance that uses out-of-bounds vertex the input values are initialized as-if they were in-bounds (`robustness2` should make them zero). Also runtime errors are thrown when OOB vertex is accessed even with `robustness2` enabled.

File: renderdoc/driver/vulkan/vk_layer.cpp (hunk 4 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_layer.cpp]: Ah OK I see what you mean. That's not really the problem case that the spec was clarifying, but you're right in that technically speaking this is wrong according to the letter of the spec although it shouldn't cause any problems in practice.

You don't need to manually annotate every function like this, it can be handled with the existing `CheckExt` macros which already annotate promotions. It just needs to be changed to not merge the check down to a single condition.

File: renderdoc/driver/vulkan/vk_core.cpp [hunk 7]
```diff
@@ -1761,8 +1765,7 @@ static const VkExtensionProperties supportedExtensions[] = {
     {
         VK_VALVE_MUTABLE_DESCRIPTOR_TYPE_EXTENSION_NAME,
         VK_VALVE_MUTABLE_DESCRIPTOR_TYPE_SPEC_VERSION,
-    },
-};
+    }};
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_core.cpp]: Please don't remove this comma, it makes the array format very weirdly.

File: renderdoc/driver/vulkan/vk_info.cpp [hunk 8]
```diff
@@ -792,6 +792,10 @@ void VulkanCreationInfo::Pipeline::Init(VulkanResourceManager *resourceMan,
                                         const VkGraphicsPipelineCreateInfo *pCreateInfo)
 {
   flags = pCreateInfo->flags;
+  auto *pPipelineFlags2 =
+      FindNextStruct(pCreateInfo, VK_STRUCTURE_TYPE_PIPELINE_CREATE_FLAGS_2_CREATE_INFO_KHR);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_info.cpp]: Please don't use `auto`, instead specify the type and do the cast at the declaration instead of below when accessing it.

File: renderdoc/driver/vulkan/vk_next_chains.cpp [hunk 9]
```diff
@@ -782,7 +782,11 @@ static void AppendModifiedChainedStruct(byte *&tempMem, VkStruct *outputStruct,
   UNWRAP_STRUCT_CAPTURE_ONLY(VK_STRUCTURE_TYPE_SEMAPHORE_GET_FD_INFO_KHR, VkSemaphoreGetFdInfoKHR,   \
                              UnwrapInPlace(out->semaphore));                                         \
   UNWRAP_STRUCT_CAPTURE_ONLY(VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR, VkSwapchainCreateInfoKHR,  \
-                             UnwrapInPlace(out->surface), UnwrapInPlace(out->oldSwapchain));
+                             UnwrapInPlace(out->surface), UnwrapInPlace(out->oldSwapchain));         \
+  COPY_STRUCT(VK_STRUCTURE_TYPE_BUFFER_USAGE_FLAGS_2_CREATE_INFO_KHR,                                \
+              VkBufferUsageFlags2CreateInfoKHR);                                                     \
+  COPY_STRUCT(VK_STRUCTURE_TYPE_PIPELINE_CREATE_FLAGS_2_CREATE_INFO_KHR,                             \
+              VkPipelineCreateFlags2CreateInfoKHR);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_next_chains.cpp]: You've not added handling for all of the structs from this extension here.

File: renderdoc/driver/vulkan/vk_serialise.cpp [hunk 10]
```diff
@@ -1445,6 +1447,18 @@ SERIALISE_VK_HANDLES();
...
+  PNEXT_STRUCT(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MAINTENANCE_5_FEATURES_KHR,                           \
+               VkPhysicalDeviceMaintenance5FeaturesKHR)                                                \
+  PNEXT_STRUCT(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MAINTENANCE_5_PROPERTIES_KHR,                         \
+               VkPhysicalDeviceMaintenance5PropertiesKHR)                                              \
+  PNEXT_STRUCT(VK_STRUCTURE_TYPE_RENDERING_AREA_INFO_KHR, VkRenderingAreaInfoKHR)                      \
+  PNEXT_STRUCT(VK_STRUCTURE_TYPE_DEVICE_IMAGE_SUBRESOURCE_INFO_KHR, VkDeviceImageSubresourceInfoKHR)   \
+  PNEXT_STRUCT(VK_STRUCTURE_TYPE_PIPELINE_CREATE_FLAGS_2_CREATE_INFO_KHR,                              \
+               VkPipelineCreateFlags2CreateInfoKHR)                                                    \
+  PNEXT_STRUCT(VK_STRUCTURE_TYPE_BUFFER_USAGE_FLAGS_2_CREATE_INFO_KHR,                                 \
+               VkBufferUsageFlags2CreateInfoKHR)                                                       \
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_serialise.cpp:1471]: There are some structs here which were promoted/aliased from `VK_EXT_image_compression_control` which are part of this extension, so they should be implemented and handled here (and removed from the EXT, as KHR outranks EXT).

File: renderdoc/driver/vulkan/vk_serialise.cpp [hunk 11]
```diff
@@ -12104,6 +12110,117 @@ void Deserialise(const VkPhysicalDeviceRayQueryFeaturesKHR &el)
...
+  SerialiseNext(ser, el.sType, el.pNext);
+
+  SERIALISE_MEMBER_OPT(pCreateInfo);
+  SERIALISE_MEMBER_OPT(pSubresource);
+}
+
+template <>
+void Deserialise(const VkDeviceImageSubresourceInfoKHR &el)
+{
+  DeserialiseNext(el.pNext);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_serialise.cpp:12469]: Optional elements (those serialised with `SERIALISE_MEMBER_OPT`) are heap-allocated with a `new` since there's no direct storage for them inside their parent struct, so they need to be deleted here.

File: renderdoc/driver/vulkan/vk_shader_cache.cpp [hunk 12]
```diff
@@ -1017,11 +1017,28 @@ void VulkanShaderCache::MakeGraphicsPipelineInfo(VkGraphicsPipelineCreateInfo &p
...
   // never create derivatives
-  ret.flags &= ~VK_PIPELINE_CREATE_DERIVATIVE_BIT;
+  flags &= ~VK_PIPELINE_CREATE_DERIVATIVE_BIT;
 
-  ret.flags &= ~VK_PIPELINE_CREATE_LIBRARY_BIT_KHR;
-  ret.flags &= ~VK_PIPELINE_CREATE_RETAIN_LINK_TIME_OPTIMIZATION_INFO_BIT_EXT;
+  flags &= ~VK_PIPELINE_CREATE_LIBRARY_BIT_KHR;
+  flags &= ~VK_PIPELINE_CREATE_RETAIN_LINK_TIME_OPTIMIZATION_INFO_BIT_EXT;
+
+  if(flags >= 0x80000000u)
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_shader_cache.cpp]: It would be nice to explicitly check `flags > VK_PIPELINE_CREATE_FLAG_BITS_MAX_ENUM` here to better document what this check is for (e.g. at first I thought this was wrong until I checked and saw that vulkan doesn't use the top bit of enums - I think to avoid signed/unsigned issues).

File: renderdoc/driver/vulkan/vk_stringise.cpp [hunk 13]
```diff
@@ -4040,3 +4041,79 @@ rdcstr DoStringise(const VkGeometryTypeKHR &el)
...
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_TRANSFORM_FEEDBACK_BUFFER_BIT_EXT);
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_TRANSFORM_FEEDBACK_COUNTER_BUFFER_BIT_EXT);
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_CONDITIONAL_RENDERING_BIT_EXT);
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR);
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_ACCELERATION_STRUCTURE_STORAGE_BIT_KHR);
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_SHADER_BINDING_TABLE_BIT_KHR);
+#ifdef VK_KHR_VIDEO_ENCODE_QUEUE
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_VIDEO_ENCODE_DST_BIT_KHR);
+    STRINGISE_BITFIELD_BIT64(VK_BUFFER_USAGE_2_VIDEO_ENCODE_SRC_BIT_KHR);
+#endif
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_stringise.cpp]: Please omit beta extensions entirely rather than using ifdefs like this which could become stale.

**Review by baldurk — CHANGES_REQUESTED** [review]

**Comment by baldurk** [comment]
As a heads up for you, once the review process has completed I'm going to keep this PR on hold until I can find the time to test this and make sure everything is working solidly.

File: renderdoc/driver/vulkan/vk_stringise.cpp (hunk 13 above)
Comment (qbojj) [reply renderdoc/driver/vulkan/vk_stringise.cpp]: Those flags should not be flagged as beta and I have submitted an issue to the Vulkan-Docs repo so this was fixed with new headers (1.3.279). I have added a comment explaining this

**Comment by baldurk** [comment]
If you want to keep prototyping or iterating on this code that's fine but can you please either close this PR or do all of the work locally, make sure you are satisfied with it, and only push when it is finalised and ready for review? Having an ever moving target is not easy to know that it's ready for review and getting email notification spam is not very pleasant either.

**Comment by qbojj** [comment]
It is just that I have found a bug in my code, so I fixed it. Other than that I am satisfied with the state of this PR

**Comment by qbojj** [comment]
I've rebased and resolved conflicts

**Comment by qbojj** [comment]
Rebased to resolve conflicts
